src/main/scala/RewriteView.scala

```diff
@@ -15,103 +18,82 @@ package fuselang
  * If `a` itself is a view, we keep rewriting it until we reach a true array.
  */
 object RewriteView {
-  import Syntax._
-  import CodeGenHelpers._
+  import Utils.State
 
-  // We can use a flat environment because the type checker makes sure things
-  // in the right scope.
   type T = Map[Id, List[Expr] => Expr]
```

Comment:
A tiny thing, but any particular reason to call this `T` instead of `Env` to suggest its purpose?

Reply:
Yeah, mostly because I wanted to extract this into a generic class.

Reply:
Is this a pattern that can be better expressed with arrows? (Not saying I’ll rewrite it into arrows)

Reply:
My apologies; I just realized I left this question hanging: I don't currently see a great way to use arrows here, but I freely admit to not being any kind of typeclass wizard. :smiley: A real Haskell person may be able to see exactly how to do it.

Reply:
Turns out there is: https://github.com/cucapra/seashell/blob/985d5d35ea97650b29fe17a2e7d17d2d475c6d4a/src/main/scala/RewriteView.scala#L38

Only took me two hours to figure out the types!

Reply:
Very nicely done!